subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/DefaultTaskProperties.java

```diff
@@ -222,7 +222,7 @@ public void visitInputProperty(String propertyName, PropertyValue value, boolean
         public void visitUnpackedOutputFileProperty(String propertyName, boolean optional, PropertyValue value, OutputFilePropertySpec spec) {
             taskPropertySpecs.add(new DefaultValidatingProperty(
                 propertyName,
-                new StaticValue(spec.getOutputFile()),
+                new ResolvingValue(value, delegate -> spec.getOutputFile()),
```

Comment:
I don't understand what the laziness improves in this case here and in `TaskExecution`. This is already the unpacked output, and for unpacking the output we already needed to get all the output files. Am I missing something?

Can you add an integration test which shows the change in behavior?

Reply:
In the current `master` we always call this method with a new `PropertyValue` object so it doesn't matter that we do this change here. But the goal is to pass the same `PropertyValue` instance here in the future. Once we do so, it will mean that by retaining the reference to this persistent `PropertyValue` via the `ResolvingValue`, once the persistent object is finalized, we can keep relying on this finalized instance.

So, essentially, two changes are needed for the desired effect, this is one of them.

Reply:
The laziness is there so we can actually gather information about all the task properties while only actually calling getters for the `@Nested` stuff.

Currently we have places where we need to get only the inputs, and not the outputs of a task (or the other way around). This mean calling only the input getters during a `PropertyWalker` walk, and not the output getters. This means for example that we can gather information about the inputs even if the output getters would throw.

What I want is to collect all the metadata about properties in a `TaskProperties` object that keeps track of a `PropertyValue` instance for each property. We can then finalize this instance and then rely on it later on. I need the lazy stuff to be able to replace the `PropertyWalker`s later on with this `TaskProperties` object with its lazy `PropertyValue`s.

Reply:
I don't understand how that would work in the future, since the call to `resolveOutputPropertySpec` is outside of the laziness here: https://github.com/gradle/gradle/pull/22594/files#diff-bd948aae1e34790fc04671cfffaf77ea9d99f7527f1065a2d24c6e3356155d11R83

If you want to make this lazy, this resolving needs be lazy, right? Am I missing something?

Reply:
Yes, we need to make some more changes in `resolveOutputPropertySpec()`. Instead of making it resolution of the properties lazy, though, I'd like to do it differently. We currently always call the getter for the output property, however, we only need the actual value to resolve plural outputs, i.e. `@OutputFiles` and `@OutputDirectories`. Most tasks use `@OuptutFile` and `@OutputDirectory`, which don't need the getter to be called early, and that's how I'd work around this. For plural outputs we'd still be calling the getters early. I was planning to do this as a next PR, but I can do it now as well if that makes it easier to follow.

Reply:
I am fine with doing this in a follow up PR. I was wondering if there is a change in behavior right now which we could show in an integration test or a unit test.

Reply:
At this point I don't expect a change in user-facing behavior. Unit test wise maybe we can add something, but all this is just a first refactor step in a series of many that I hope will substantially reduce complexity. For now I think the existing coverage is enough.